src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,684 @@
+#include "xlog.h"
+#include "memtx_sort_data.h"
+#include "memtx_index_read_view.h"
+
+#include "core/assoc.h"
+#include "trivia/util.h"
+
+#define DIAG_SET(msd, ...) do { \
+	diag_set(ClientError, ER_INVALID_SORTDATA_FILE, \
+		 (msd)->fname, tt_sprintf(__VA_ARGS__)); \
+} while (false)
+
+/**
+ * Map: (space_id, index_id) => (sort data entry information).
+ */
+struct memtx_sort_data_key {
+	/** The space ID. */
+	uint32_t space_id;
+	/** The index ID. */
+	uint32_t index_id;
+};
+
+static inline uint64_t
+memtx_sort_data_key_hash(struct memtx_sort_data_key key)
+{
+	return (key.space_id << 8) | key.index_id;
+}
+
+static inline int
+memtx_sort_data_key_cmp(struct memtx_sort_data_key a,
+			struct memtx_sort_data_key b)
+{
+	return a.space_id != b.space_id || a.index_id != b.index_id;
+}
+
+#define mh_name _memtx_sort_data_entries
+#define mh_key_t struct memtx_sort_data_key
+
+/** The sort data file header entry (the index sort data information). */
+struct memtx_sort_data_entry {
+	/** The entry identifier. */
+	struct memtx_sort_data_key key;
+	/** The offset of the sort data in the file. */
+	long offset;
+	/** The physical size of the sort data. */
+	long psize;
+	/** The amount of tuples stored. Can be updated by reader in process. */
+	long len;
+	/** Writer-only: offset of offset field in the file header. */
+	long offset_offset;
+	/** Writer-only: offset of psize field in the file header. */
+	long psize_offset;
+	/** Writer-only: offset of len field in the file header. */
+	long len_offset;
+	/** Writer-only: the data is fully written into the file already. */
+	bool is_committed;
+};
+
+#define mh_node_t struct memtx_sort_data_entry
+#define mh_arg_t void *
+#define mh_hash(a, arg) (memtx_sort_data_key_hash((a)->key))
+#define mh_hash_key(a, arg) (memtx_sort_data_key_hash(a))
+#define mh_cmp(a, b, arg) (memtx_sort_data_key_cmp((a)->key, (b)->key))
+#define mh_cmp_key(a, b, arg) (memtx_sort_data_key_cmp(a, (b)->key))
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+/** The sort data file reader and writer context. */
+struct memtx_sort_data {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The XLOG file directory. */
+	struct xdir dir;
```

Comment:
xdir/xlog shouldn't be used for reading/writing sortdata files because they have a completely different format.

Reply:
There's no need to scan the directory for `.sortdata` files. It's safe to assume that a `.sortdata` file can only exist if there's a `.snap` file with the same signature. The garbage collector should first remove `.sortdata` files, then `.snap` files.

As for file format validation, it's trivial to write and parse kv strings in text format. I don't see much point in reusing `xlog` for that.

Reply:
Regarding the garbage collection, see the [comment above](https://github.com/tarantool/tarantool/pull/11133#discussion_r2052049629).

Made the sort data file of its own format.